src/common/src/com/microsoftopentechnologies/tooling/msservices/helpers/auth/AADManagerImpl.java

```diff
@@ -0,0 +1,697 @@
+        }
+
+        return wrappedFuture;
+    }
+
+    private <T> T authenticatedRequest(@NotNull final UserInfo userInfo,
+                                       @NotNull final String resource,
+                                       @NotNull final String title,
+                                       @NotNull final RequestCallback<T> requestCallback)
+            throws AzureCmdException {
+        AuthenticationResult authenticationResult = getAuthenticationResult(userInfo, resource);
+
+        try {
+            return requestCallback.execute(authenticationResult.getAccessToken());
+        } catch (Throwable throwable) {
+            if (isErrorResourceUnauthorized(throwable)) {
+                return unauthenticatedRequest(userInfo, resource, title, requestCallback);
+            } else if (throwable instanceof AzureCmdException) {
+                throw (AzureCmdException) throwable;
+            } else if (throwable instanceof ExecutionException) {
+                throw new AzureCmdException(throwable.getCause().getMessage(), throwable.getCause());
+            }
+
+            throw new AzureCmdException(throwable.getMessage(), throwable);
+        }
+    }
+
+    private <V> void authenticatedRequest(@NotNull final UserInfo userInfo,
+                                          @NotNull final String resource,
+                                          @NotNull final String title,
+                                          @NotNull final RequestCallback<ListenableFuture<V>> requestCallback,
+                                          final SettableFuture<V> wrappedFuture) {
+        try {
+            AuthenticationResult authenticationResult = getAuthenticationResult(userInfo, resource);
+
+            Futures.addCallback(requestCallback.execute(authenticationResult.getAccessToken()), new FutureCallback<V>() {
+                @Override
+                public void onSuccess(V val) {
+                    wrappedFuture.set(val);
+                }
+
+                @Override
+                public void onFailure(Throwable throwable) {
+                    if (isErrorResourceUnauthorized(throwable)) {
+                        unauthenticatedRequest(userInfo, resource, title, requestCallback, wrappedFuture);
+                    } else {
+                        wrappedFuture.setException(throwable);
+                    }
+                }
+            });
+        } catch (Throwable e) {
+            wrappedFuture.setException(e);
+        }
+    }
+
+    private <T> T unauthenticatedRequest(@NotNull UserInfo userInfo,
+                                         @NotNull String resource,
+                                         @NotNull String title,
+                                         @NotNull RequestCallback<T> requestCallback)
+            throws AzureCmdException {
+        if (hasRefreshAuthResult(userInfo)) {
+            authenticateWithRefreshToken(userInfo, resource, title);
```

Comment:
As before, maybe we can combine the check for presence of refresh tokens with its use via null checks or `Optional<T>`?